app/src/main/java/com/sunasterisk/travelapp/data/LoadDataAsync.kt

```diff
@@ -2,27 +2,23 @@ package com.sunasterisk.travelapp.data
 
 import android.os.AsyncTask
 
-class LoadDataAsync<T, P>(
+class LoadDataAsync<T, P : Any>(
     private val callbackListener: OnDataCallback<P>,
-    private val handler: (T) -> Any
-) : AsyncTask<T, Unit, Any>() {
-    override fun doInBackground(vararg params: T): Any =
+    private val handler: (T) -> Result<P>?
+) : AsyncTask<T, Unit, Result<P>?>() {
+    override fun doInBackground(vararg params: T): Result<P>? =
         try {
             handler(params[0])
         } catch (exception: Exception) {
-            exception
+            Result.Error(exception)
         }
 
```

Comment:
Không thể mong đợi exception được mà return exception 
sử dụng biến global lưu exception lại

Reply:
Em  để kết quả trả về của doInBackground là  1 object Result nếu có exception thì result là Error  xong ở onPostExecute  sẽ xử lý result đó. E không hiểu tại sao lại cần lưu exception bằng biến global

Reply:
vì AsyncTask là thực hiện 1 task nào cả đồng bộ
và chả ai thực hiện 1 task trả về exception cả

Reply:
`
class LoadDataAsync<T, P>(
    private val callbackListener: OnDataCallback<P>,
    private val handler: (T) -> Result<P>?
) : AsyncTask<T, Unit, Result<P>?>() {
    override fun doInBackground(vararg params: T): Result<P>? =
        try {
            handler(params[0])
        } catch (exception: Exception) {
          //  Result.Error(exception)
        }

    override fun onPostExecute(result: Result<P>?) {
        super.onPostExecute(result)
        when(result){
            is Result.Error ->    callbackListener.onError(result.exception)
            is Result.Success ->  callbackListener.onSuccess(result.data)
            else -> callbackListener.onError(Exception(EXCEPTION))
        }
    }

    companion object {
        const val EXCEPTION = "Loading data failed, exception occurred"
    }
}
`